packages/block-library/src/accordion-item/index.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/**
+ * Server-side rendering of the `core/accordion-item` block.
+ *
+ * @package WordPress
+ * @since 6.6.0
+ *
+ * @param array $attributes The block attributes.
+ * @param string $content The block content.
+ *
+ * @return string Returns the updated markup.
+ */
+function block_core_accordion_item_render( $attributes, $content ) {
+	if ( ! $content ) {
+		return $content;
+	}
+
+	$p         = new WP_HTML_Tag_Processor( $content );
+	$unique_id = wp_unique_id( 'accordion-item-' );
+
+	while ( $p->next_tag() ) {
+		if ( $p->has_class( 'wp-block-accordion-item' ) ) {
+            $p->set_attribute( 'data-wp-context', '{ "id": "' . $unique_id . '" }' );
+			$p->set_attribute( 'data-wp-class--is-open', 'state.isOpen' );
+			if ( $attributes['openByDefault'] ) {
+				$p->set_attribute( 'data-wp-init', 'callbacks.open' );
+			}
```

Comment:
As you are using JavaScript to open the item, the initial HTML sent to the browser will have the item closed and will only open after the JavaScript loads, causing a layout shift. 

It would be better if the state on the server could already contain the correct information about which item is opened so that the HTML the browser receives is the correct HTML.

There are several ways to share information between blocks. One option could be something similar to what [we've done in the image block](https://github.com/WordPress/gutenberg/pull/63348), where we put the info in the global state and only an `id` in the local context. But it's not the only way, so feel free to choose the best one for you.

_(By the way, I am preparing a new guide on server-side rendering. Once I open the pull request, I will put a link here.)_

Reply:
> It would be better if the state on the server could already contain the correct information about which item is opened so that the HTML the browser receives is the correct HTML.

I tried to initialize the state on the server: https://github.com/WordPress/gutenberg/pull/64119/commits/4c27428623c99430d65d4cf7ade64a233a9dca14

There are two issues now:

1. I'm still getting layout shift, I don't know why this is the case since I thought the state is being initialized with the right info on the server.
2. It breaks if there is more than one accordion group on the page and `autoclose` (only allows one item per group to be open at a time) is enabled, since they are sharing the same store and I'm using global state to track which ids are open.

Do you have any suggestions? Thank you for the review so far.

Reply:
> I'm still getting layout shift, I don't know why this is the case since I thought the state is being initialized with the right info on the server.

The derived state was missing on the server. 

> It breaks if there is more than one accordion group on the page and autoclose (only allows one item per group to be open at a time) is enabled, since they are sharing the same store and I'm using global state to track which ids are open.

To fix the autoclose issue, I refactored the code to move the array with the IDs of open items to the local context.

I've made a video to explain the two changes.

<div>
    <a href="https://www.loom.com/share/8b44e78546b7415bba02de825f59c045">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/8b44e78546b7415bba02de825f59c045-ccb230eecab2ed0e-full-play.gif">
    </a>
  </div>

https://www.loom.com/share/8b44e78546b7415bba02de825f59c045